pkg/datastore/headers.go

```diff
@@ -0,0 +1,433 @@
+
+	if el == nil {
+		el = []structs.HeaderAndTrace{}
+	}
+
+	// all from memory
+	if len(el) >= int(limit) {
+		return el[:limit], nil
+	}
+
+	initialSlot := lastSlot
+	readr := bytes.NewReader(nil)
+	dec := json.NewDecoder(readr)
+	for {
+		data, err := s.TTLStorage.Get(ctx, HeaderKey(initialSlot))
+		if err != nil {
+			if errors.Is(err, ds.ErrNotFound) {
+				return el, nil
+			}
+			return el, err
+		}
+		readr.Reset(data)
+		hnt := []structs.HeaderAndTrace{}
+		if err := dec.Decode(&hnt); err != nil {
+			return nil, err
+		}
+
+		el = append(el, hnt...)
+		initialSlot--
+		// introduce limit?
+	}
+}
+
+func (s *Datastore) deprecatedGetHeaders(ctx context.Context, key ds.Key) ([]structs.HeaderAndTrace, error) {
+	data, err := s.TTLStorage.Get(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+	return s.deprecatedUnsmarshalHeaders(data)
+}
+
+func (s *Datastore) deprecatedUnsmarshalHeaders(data []byte) ([]structs.HeaderAndTrace, error) {
+	var headers []structs.HeaderAndTrace
+	if err := json.Unmarshal(data, &headers); err != nil {
+		var header structs.HeaderAndTrace
+		if err := json.Unmarshal(data, &header); err != nil {
+			return nil, err
+		}
+		return []structs.HeaderAndTrace{header}, nil
+	}
+	return headers, nil
+}
+
+// SaveHeaders is meant to persist the all the keys under one key
+// As optimization in future this function can operate only on database, so instead from memory it may just reorganize keys
+func (s *Datastore) SaveHeaders(ctx context.Context, slots []uint64, ttl time.Duration) error {
+	for _, slot := range slots {
+		if err := s.saveHeader(ctx, slot, ttl); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Datastore) saveHeader(ctx context.Context, slot uint64, ttl time.Duration) error {
+	el, rev, err := s.hc.GetSingleSlot(slot)
+	if err != nil {
+		return err
+	}
+
+	if err = putHeaders(ctx, s, slot, el, ttl); err != nil {
+		return err
+	}
+
+	if s.hc.UnlinkElement(slot, rev) {
+		return nil // success
+	}
+
+	// revert the saveHeaders operation as revision changed
+	_ = s.Badger.Update(func(txn *badger.Txn) error {
+		return txn.Delete(HeaderKey(slot).Bytes())
+	})
+
+	return errors.New("revision changed")
+}
+
+func putHeaders(ctx context.Context, s *Datastore, slot uint64, cont []structs.HeaderAndTrace, ttl time.Duration) error {
+	buff := bytes.NewBuffer(nil)
+	buff.WriteString("[")
+
+	enc := json.NewEncoder(buff)
+
+	for i, c := range cont {
+		if i > 0 {
+			buff.WriteString(",")
+		}
+		if err := enc.Encode(c); err != nil {
+			return err
+		}
+	}
+	buff.WriteString("]")
+	if err := s.TTLStorage.PutWithTTL(ctx, HeaderKey(slot), buff.Bytes(), ttl); err != nil {
+		return err
+	}
+
+	buff.Truncate(0) // immediately remove
+	return nil
+}
+
+type LoadItem struct {
+	Time    uint64
+	Content []byte
+}
+
+// FixOrphanHeaders is reading all the orphan headers from
+func (s *Datastore) FixOrphanHeaders(ctx context.Context, ttl time.Duration) error {
```

Comment:
We should add unit tests for this function. Since some time ago Badgerdb had some issues/bugs for iterating based on a prefix. But probably the bug is fixed